### PR TITLE
bibcheck: new rule: fix_ref_jhep_volume

### PR DESCRIPTION
**Pull request by ksachs**
* try to fix 999C5s for JCAP, JHEP if volume is a year
* includes main() function to run stand-alone
* not sure if this should run weekly (one ticket/rule) or standard (one
ticket/record)

Signed-off-by: Kirsten Sachs <sachs@l00lnxkaos.desy.de>

### CODE REVIEW

**Comment by ksachs** [comment]
For now this is only for JCAP, which I cleaned already offline.
I'll change the rule to include JHEP and J.Stat.Mech. as soon as they are clean too.

Open question:
should this be part of weekly-rules.cfg (one ticket / rule) or rules.cfg (one ticket / record).
I have to idea how many records will be usually flagged to be fixed manually.
For a larger number I would prefer to process them in batch (ticket/rule).
On the other hand it would be better to run the check more frequently to get the automatic fixes in.
My idea is to let it run weekly for now.

File: bibcheck/weekly-rules.cfg [hunk 1]
```diff
@@ -1,3 +1,8 @@
 [KAOS_journal_names]
 check = journal_names
 filter_collection = Journals
+
+[KAOS_fix_ref.JCAP_JHEP]
+check = fix_ref_jhep_volume.py
```
Comment (tsgit) [inline bibcheck/weekly-rules.cfg]: no `.py` in rules, this errors
> 2019-09-16 12:07:22 --> Invalid rule 'KAOS_fix_ref.JCAP_JHEP': Invalid checker 'fix_ref_jhep_volume.py'.

File: bibcheck/plugins/fix_ref_jhep_volume.py [hunk 2]
```diff
@@ -0,0 +1,490 @@
...
+            # this looks like number of pages, not page-numbers
+            artid = ''
+        if page_range[0] == page_range[1]:
+            # this looks like artid-artid
+            artid = page_range[0]
+    if 0 < len(artid) < 3:
+        # add leading 0s
+        try:
+            artid = '%03i' % int(artid)
+        except:
```
Comment (tsgit) [inline bibcheck/plugins/fix_ref_jhep_volume.py]: `except ValueError:`

File: bibcheck/plugins/fix_ref_jhep_volume.py [hunk 3]
```diff
@@ -0,0 +1,490 @@
...
+                    reference_dict['position_pbn'] = subfield_position
+                    reference_dict['subfield_pbn'] = value
+
+    return reference_dict
+                   
+def parse_pbn(reference):
+    """ split $$s subfield in journal, volume and artid """
+    try:
+        (journal, volume, artid) = reference['subfield_pbn'].split(',')
+    except:
```
Comment (tsgit) [inline bibcheck/plugins/fix_ref_jhep_volume.py]: `except (KeyError, ValueError):`

**Comment by tsgit** [comment]
I attached the bibcheck log from a test run.

It modified 355 records, e.g.

```
     <subfield code="h">Cline, J.M.</subfield>
     <subfield code="h">Kainulainen, K.</subfield>
     <subfield code="o">17</subfield>
-    <subfield code="s">JCAP,2013,012</subfield>
+    <subfield code="s">JCAP,1301,012</subfield>
     <subfield code="y">2020</subfield>
     <subfield code="z">0</subfield>
   </datafield>
@@ -718,7 +753,7 @@
     <subfield code="a">doi:10.1088/1475-7516/2016/04/001</subfield>
     <subfield code="h">Caprini, C.</subfield>
     <subfield code="o">41</subfield>
-    <subfield code="s">JCAP,2016,001</subfield>
+    <subfield code="s">JCAP,1604,001</subfield>
     <subfield code="y">2020</subfield>
     <subfield code="z">0</subfield>
   </datafield>
```
some more randomly picked examples

-    <subfield code="s">JCAP,2017,047</subfield>
+    <subfield code="s">JCAP,1705,047</subfield>

-    <subfield code="s">JCAP,2016,023</subfield>
+    <subfield code="s">JCAP,1607,023</subfield>

-    <subfield code="s">JCAP,2018,009-009</subfield>
+    <subfield code="s">JCAP,1803,009</subfield>

-    <subfield code="s">JCAP,2011,040</subfield>
+    <subfield code="s">JCAP,1103,040</subfield>

-    <subfield code="s">JHEP,2012,009</subfield>
+    <subfield code="s">JHEP,1209,009</subfield>

-    <subfield code="s">JHEP,2012,054</subfield>
+    <subfield code="s">JHEP,1201,054</subfield>

-    <subfield code="s">JCAP,2014,056</subfield>
+    <subfield code="s">JCAP,1408,056</subfield>

I guess that's all good

File: bibcheck/weekly-rules.cfg (hunk 1 above)
Comment (ksachs) [reply bibcheck/weekly-rules.cfg]: sorry

**Comment by ksachs** [comment]
I must admit the bare exceptions are on purpose. I was too lazy to be sure there is no other exception possible given the fact that you never know what we have in the metadata. Main purpose is to prevent the code from crashing - no matter what.

**Comment by ksachs** [comment]
I added an option to run without creating tickets. If we run it in the normal bibcheck (one ticket per record) I doubt we can find the tickets. So I'll have to parse the log-file.

**Pull request closed without merging**